bundles/automation/org.eclipse.smarthome.automation.api/src/main/java/org/eclipse/smarthome/automation/handler/BaseModuleHandlerFactory.java

```diff
@@ -52,12 +54,11 @@ public void dispose() {
 
     @Override
     public void ungetHandler(Module module, String ruleUID, ModuleHandler hdlr) {
-        ModuleHandler handler = handlers.get(ruleUID + module.getId());
+        hdlr.dispose();
```

Comment:
This should not be done until it is the last reference of this handler! As HanlderFactories can decide to reuse instances of handlers for other modules it might happen that the same instance is kept with different keys in this map!